File: bench/equivalents/python/fib.py
#!/usr/bin/env python3
# Python equivalent of bench/fib.ir: naive recursive Fibonacci.
#
# Invoked by bench/bench.tcl for the interp/compile/python/rust comparison
# table. Prints a "value:"/"best_us:" pair for the caller to parse; n comes
# from argv (default matches the literal in fib.ir) so nothing here is a
# compile-time constant to an optimizing runtime.
import sys
import time


def fib(n):
    if n < 2:
        return n
    return fib(n - 1) + fib(n - 2)


def parse_args(argv):
    runs = 5
    positional = []
    i = 0
    while i < len(argv):
        if argv[i] == "--runs":
            i += 1
            runs = int(argv[i])
        else:
            positional.append(argv[i])
        i += 1
    return runs, positional


if __name__ == "__main__":
    runs, positional = parse_args(sys.argv[1:])
    n = int(positional[0]) if positional else 28

    fib(n)  # untimed warmup
    best = None
    value = None
    for _ in range(runs):
        t0 = time.perf_counter()
        value = fib(n)
        dt = (time.perf_counter() - t0) * 1_000_000
        if best is None or dt < best:
            best = dt

    print(f"value: {value}")
    print(f"best_us: {best:.3f}")

File: bench/equivalents/rust/fib.rs
// Rust equivalent of bench/fib.ir: naive recursive Fibonacci.
//
// Invoked by bench/bench.tcl for the interp/compile/python/rust comparison
// table. Prints a "value:"/"best_us:" pair for the caller to parse; n comes
// from argv (default matches the literal in fib.ir) so rustc never sees a
// compile-time constant to fold.
use std::hint::black_box;
use std::time::Instant;

fn fib(n: i64) -> i64 {
    if n < 2 {
        n
    } else {
        fib(n - 1) + fib(n - 2)
    }
}

fn main() {
    let mut runs: u32 = 5;
    let mut positional = Vec::new();
    let mut args = std::env::args().skip(1);
    while let Some(arg) = args.next() {
        if arg == "--runs" {
            runs = args.next().expect("--runs needs a value").parse().expect("--runs must be an integer");
        } else {
            positional.push(arg);
        }
    }
    let n: i64 = positional.get(0).map(|s| s.parse().expect("n must be an integer")).unwrap_or(28);

    fib(black_box(n)); // untimed warmup
    let mut best = f64::MAX;
    let mut value = 0;
    for _ in 0..runs {
        let t0 = Instant::now();
        // black_box on the argument stops the optimizer from noticing the
        // call is loop-invariant and hoisting it out after the first run.
        value = fib(black_box(n));
        let us = t0.elapsed().as_secs_f64() * 1_000_000.0;
        if us < best {
            best = us;
        }
    }

    println!("value: {}", value);
    println!("best_us: {:.3}", best);
}
